### PR TITLE
New version 3

### PR DESCRIPTION
**Pull request by Worvast**
#### Added 
 * Faker
     * Faker time periods for probability and frequency
     * SimulationGenerator
     * New providers
     * New and splitted documentation of use
     * Examples for scripts and CLI shell
     * Generators/Faker now accept custom providers, custom functions 
     from base python or custom scripts for use in templates
     * CLI Shell has now all necessary flags for use all generators 
 
#### Changed
 * General
    * Modified base devo-sender from version 3.0.x to 3.3.0
    * Update dependencies to a new majors versions
    * Set dependencies to fixed versions (more maintenance, but much more security and reliability)
 * Sorter
    * file_sorted_join moved from devoutils.sorter to devoutils.fileio
    * Remove Python 2 compatibility
 * Faker
    * Faker cli add verbose mode to show the events in the console
    * Add file_name param to define a file to store events in batch mode or 
     for testing, store in a file but do not send it.
    * Generator template are moved for Template
    * Generator date_generator are moved for Template
    * Generators name changed, example: BatchSender for BatchFakerGenerator
    * Now you can make functions callable in each line of Faker Jinja2 Template
    * null_date_generator (Default date generator) isn't a generator now, its a normal function
    * Template.process() now not create new DateGenerator in each call
    * Numbers_providers has PEP8 variable names
    * `freq` var/flag in all code are now `frequency`
    * `prob` var/flag in all code are now `probability`
    * CLI Shell --config file now require all vars in "faker" object, and all vars for Sender in "sender" object

### CODE REVIEW

**Comment by Worvast** [comment]
I am not sure if I have put everything in the changelog, if you see more undocumented changes warn

File: devoutils/faker/scripts/faker_cli.py [hunk 1]
```diff
@@ -32,21 +33,30 @@ def cli():
...
 @click.option('--interactive', '-i', is_flag=True,
               help='Interactive mode.')
 @click.option('--raw_mode', '-raw', is_flag=True,
               help='Send raw mode.')
-@click.option('--prob', default=100, help='Probability (0-100).')
-@click.option('--freq', default="1-1", help='Frequency in seconds. Example:'
-                                            '"1.0-5.0" = random time '
-                                            'between 1 sec. to 5secs.')
+@click.option('--probability', default=100, help='Probability (0-100).')
+@click.option('--frequency', default="1-1", help='Frequency in seconds. '
```
Comment (aarteagamel) [inline devoutils/faker/scripts/faker_cli.py:52]: I think that the 99% of the use of the faker is with cli. I haven't seen the option to use an "Complex probability and frequency", If there is no option for this I think that we must add it

File: docs/faker.md [hunk 2]
```diff
@@ -144,38 +152,93 @@ class TemplateParser:
...
+
+Faker will not treat that line as a send line, if a variable is created without the hyphens, for example:
 
+    {% set .... %}
 
-## TODO
-- Add more useful methods to the jinja collection
-- Documentate news methods in Faker
-- Provide use examples in script
+Faker will believe that it is a line to send, so you have to keep this in mind
```
Comment (aarteagamel) [inline docs/faker.md:243]: I don't understand this comment

File: docs/faker/batch.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+The batch mode allows to generate events between two arbitrary dates as fast 
+as possible. The events will be dumped to the standard output and then you can
+redirect it to a file or somewhere else
+
+First we want to take the sample events and create a template from them. For
+example, if we have these sample events:
+
+```
+<14>2018-03-22T23:56:09.237-07:00 ABCDEFGHI01 firewall.cisco_asa: 2018-03-22T23:56:09.237067-07:00 : %ASA-123456: Teardown dynamic UDP translation from inside:123.123.123.123/00001 to outside:111.111.111.111/11111 duration 0:02:32
```
Comment (aarteagamel) [inline docs/faker/batch.md:21]: The batch mode requires the events in raw format? if true we must put it in the documentation

File: docs/faker/file.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+## File fake generator
```
Comment (aarteagamel) [inline docs/faker/file.md:1]: Do we have file fake generator allowed in the cli. If not I think is a good idea to add it and document

File: docs/faker/realtime.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+## Realtime fake generator
```
Comment (aarteagamel) [inline docs/faker/realtime.md:1]: If we have cli options for realtime fake generator add an example. Maybe is int the cli documentation.

File: docs/faker/shellcli.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                               removed from the generated dates by doing
+                               date[:-3] this flags prevents it
+
+  --verbose                    Verbose mode shows the events created in the
+                               console when sending dta into Devo.
+
+  --help                       Show this message and exit.
+```
+
+The config.json file could be like this:
```
Comment (aarteagamel) [inline docs/faker/shellcli.md:68]: I miss a list with all the values allowed. If the config refers to SDK, then link to the SDK documentation but right now we only have an example

File: docs/faker/simulation.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+## Simulation fake generator
```
Comment (aarteagamel) [inline docs/faker/simulation.md:1]: Add cli example

File: docs/faker/syslog.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
+## Syslog fake generator
```
Comment (aarteagamel) [inline docs/faker/syslog.md:1]: add cli example

File: examples/cisco_asa_cli_example.jinja2 [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+{%- set protocol = fake.random_element(('TCP', 'UDP')) -%}
+{%- set outside_ip = fake.random_element(("94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "4.2.23.1", "4.2.23.12", "4.2.23.13")) -%}
+{%- set outside_port = fake.int(1, 65000) -%}
+{%- set inside_ip = fake.random_element(("94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "94.2.23.1", "4.2.23.1", "4.2.23.12", "4.2.23.13")) -%}
+{%- set inside_port = fake.int(1, 65000) -%}
+{#- Inbound connections -#}
+<14>{{ eventdate }} injection-eu-52.19.12.121 firewall.cisco.asa: 2018-03-22T23:56:09.076432-07:00 TESTEWHQMHF28 : {{ ASA_ids }}: Built inbound {{ protocol }} connection 684813185 for outside:{{ outside_ip }}/{{ outside_port }} ({{ outside_ip }}/{{ outside_port }})(LOCAL\BLAH, Sara) to inside:{{ inside_ip }}/{{ inside_port }} ({{ inside_ip }}/{{ inside_port }}) (Mostafavi, Sara)
```
Comment (aarteagamel) [inline examples/cisco_asa_cli_example.jinja2:12]: instead of sara use something like Fake Name

File: examples/dns-bind-query_cli_example.jinja2 [hunk 10]
```diff
@@ -0,0 +1,40 @@
+{{ fake.date_time_between(start_date='-2s', end_date='now', tzinfo=None).strftime("%d-%b-%Y %H:%M:%S.%f")[:-3] }} queries: info: client {{ fake.ipv4() }}#50440: query: {{ fake.domain_name() }} IN A +
```
Comment (aarteagamel) [inline examples/dns-bind-query_cli_example.jinja2:1]: Use private Ips

File: devoutils/faker/scripts/faker_cli.py (hunk 1 above)
Comment (Worvast) [reply devoutils/faker/scripts/faker_cli.py:52]: I put it in all function but the documentation say, if hard to put complex frequency and probability in --flags, soo, I only put if you add it in the config.json/.yaml, not a a flag. If you have one idea about how we can manage in a shell the complex probability and frequency, send a message please!

**Pull request merged**